modules/internal/CPtr.chpl

```diff
@@ -377,6 +377,31 @@ module CPtr {
     return sizeof(x);
   }
 
+  /*
+    Return the offset of elements in a record.
+
+    .. warning::
+
+      This method is intended for C interoperability.  To enhance flexibility,
+      it is possible to request the offset of elements within a Chapel record.
+      However, be aware:
+
+      * Chapel types are not necessary stored in contiguous memory
+      * Behavior of ``c_offsetof`` may change
+      * Behavior given a Chapel class type field is not well-defined
+   */
+  proc c_offsetof(type t, param fieldname : string) where isRecordType(t) {
+    use Reflection;
+    var x: t;
+
+    return c_ptrTo(getFieldRef(x, fieldname)):uint - c_ptrTo(x):uint;
```

Comment:
I have some suggestions:

1. I think offsetof should apply to class members just as well as record members.
2. This function constructs and then destroys a record. That's not necessary however since we could just make `nil` C pointer to the record and the get the address of the field.
3. I'm tempted to have it call the C macro offsetof, to match sizeof. But, note that such a thing would require us to adjust the LLVM code generator since offsetof will need to be handled by it directly. In that way the current approach is easier.

Reply:
One possible advantage of using the C macro is that it would remove the need to `use Reflection` here.

Reply:
`nil` can't be cast to a record type, it's for classes only.  I don't think there's something that can be used in the case of records, since they don't all inherit from the same place.

Would the usage of the offset vary depending on if we were looking at a class or a record?  We would need to dereference the class pointer first, right?

Reply:
Trying to use `nil` for classes also results in complaints about dereferencing nil, which might indicate we need to either create or as for a class instance

Reply:
Probably ask for one, since we can't rely on a particular initializer being present for the type.  It seems like it would be weird to require a type argument in the record case and a type instance in the class case, though.  Does that mean that we should require an instance for both?

Reply:
@lydia-duncan - the `nil` thing isn't the example variable. Rather, you create a `c_ptr` that is `nil` that points to the class instance or to the record. For example:

``` chapel
var nilPtr: c_ptr(t) = nil;
ref fieldRef = nilPtr.deref().field;
var eltPtr = c_ptrTo(fieldRef);
return eltPtr: size_t;
```

I havn't tried this. My one concern about this pattern in particular is that everywhere else we can assume a `ref` is not `nil`, but here it is. But we don't dereference it, so it probably doesn't matter.

Reply:
no, it computes the address of it and stores it in a reference

Reply:
does it work if you hardcode the field? getFieldRef shouldn't be dereferencing it since it accepts its argument by ref.

Reply:
Hardcoding it gives me a segfault due to a bad access (which I think the nil dereference check was protecting me against):

```chapel
var nilPtr: c_ptr(t) = nil;
ref fieldRef = nilPtr.deref().x; // This line complains
var eltPtr = c_ptrTo(fieldRef);
writeln( eltPtr: size_t - nilPtr: size_t);
```
generates this C code:
```
   15  	  nilPtr_chpl = ((c_ptr_Foo)(nil));
   16  	  call_tmp_chpl2 = (nilPtr_chpl + INT64(0));
-> 17  	  coerce_tmp_chpl = *(call_tmp_chpl2);
   18  	  chpl_check_nil(coerce_tmp_chpl, INT64(7), INT32(57));
   19  	  call_tmp_chpl = &((coerce_tmp_chpl)->x_chpl);
   20  	  fieldRef_chpl = call_tmp_chpl;
```

Reply:
You might not choose to implement it this way, but I think that's a bug. Nothing about the pattern here should have led to a dereference in the C code.